src/NumericalAlgorithms/Spectral/Spectral.tpp

```diff
@@ -0,0 +1,259 @@
+  const DataVector& collocation_pts =
+      collocation_points_and_weights<BasisType>(number_of_pts)
+          .collocation_pts();
+  const DataVector& bary_weights =
+      collocation_points_and_weights<BasisType>(number_of_pts)
+          .barycentric_weights();
+
+  Matrix diff_matrix(number_of_pts, number_of_pts);
+  for (size_t i = 0; i < number_of_pts; ++i) {
+    double diagonal = 0.0;
+    for (size_t j = 0; j < number_of_pts; ++j) {
+      if (LIKELY(i != j)) {
+        diff_matrix(i, j) =
+            bary_weights[j] /
+            (bary_weights[i] * (collocation_pts[i] - collocation_pts[j]));
+        diagonal -= diff_matrix(i, j);
+      }
+    }
+    diff_matrix(i, i) = diagonal;
+  }
+  return diff_matrix;
+}
+
+template <typename BasisType>
+Matrix compute_spectral_to_grid_points_matrix(
+    const size_t number_of_pts) noexcept {
+  // Since u(x) = sum a_k P_k(x), matrix for u(x_j) is just P_k(x_j) transposed
+  const DataVector& collocation_pts =
+      collocation_points_and_weights<BasisType>(number_of_pts)
+          .collocation_pts();
+  Matrix spec_to_grid(number_of_pts, number_of_pts);
+  // number_of_pts >=2 ASSERT'd elsewhere
+  for (size_t j = 0; j < number_of_pts; ++j) {
+    spec_to_grid(j, 0) = 1.0;  // P_0 = 1
+    double p_n_minus2 = 1.0;
+    double p_n_minus1 = collocation_pts[j];  // P_1 = x
+    spec_to_grid(j, 1) = p_n_minus1;
+    for (size_t k = 2; k < number_of_pts; ++k) {
+      const double p_n = ((2 * k - 1) * collocation_pts[j] * p_n_minus1 -
+                          (k - 1) * p_n_minus2) /
+                         k;
+      spec_to_grid(j, k) = p_n;
+      p_n_minus2 = p_n_minus1;
+      p_n_minus1 = p_n;
+    }
+  }
+  return spec_to_grid;
+}
+
+// GridPointsToSpectral is the inverse matrix to above. Can compute numerically
+// or use analytic expressions P_k(x_j) w_j/gamma_k where w_j=weight,
+// gamma_k = normalization = 2/(2k+1), except = 2/N for k=N (for GLL)
+template <typename BasisType>
+Matrix compute_grid_points_to_spectral_matrix(
+    const size_t number_of_pts) noexcept {
+  const DataVector& x = collocation_points_and_weights<BasisType>(number_of_pts)
+                            .collocation_pts();
+  const DataVector& w = collocation_points_and_weights<BasisType>(number_of_pts)
+                            .quadrature_weights();
+  Matrix g_to_s_matrix(number_of_pts, number_of_pts);
+  for (size_t j = 0; j < number_of_pts; ++j) {
+    g_to_s_matrix(0, j) = 0.5 * w[j];  // P_0 = 1
+    // number_of_pts >=2 ASSERT'd elsewhere
+    double p_n_minus2 = 1.0;
+    double p_n_minus1 = x[j];  // P_1 = x
+    g_to_s_matrix(1, j) =
+        p_n_minus1 * w[j] * ((number_of_pts == 2) ? 0.5 : 1.5);
+    for (size_t k = 2; k < number_of_pts; ++k) {
+      const double p_n =
+          ((2 * k - 1) * x[j] * p_n_minus1 - (k - 1) * p_n_minus2) / k;
+      const double gamma = (k == number_of_pts - 1)
+                               ? 2.0 / (number_of_pts - 1.0)
```

Comment:
I think the special case when gamma = 2.0/(number_of_pts-1.0) is only for GLL, not GL